smaily-for-wp.php

```diff
@@ -89,198 +90,38 @@ function smaily_admin_render() {
-		wp_safe_redirect( $redirect_url );
-		return;
-	}
-
-	// Form data required.
-	if ( ! $_POST ) {
-		$redirect_url = add_query_arg( 'smaily_form_error',
-			rawurlencode( __( 'No form data.', 'wp_smaily' ) ), $redirect_url );
-		wp_safe_redirect( $redirect_url );
-		exit;
-	}
-
-	// Parse form data out of POST and sanitize.
-	$params = array();
-	foreach( $_POST as $key => $value ) {
-		if ( $key === "action" || $key === "nonce" ) {
-			continue;
-		}
-		$params[ $key ] = sanitize_text_field( $value );
-	}
-
-	// E-mail required.
-	if ( ! ( isset( $params['email'] ) && ! empty( $params['email'] ) ) ) {
-		$redirect_url = add_query_arg( 'smaily_form_error',
-			rawurlencode( __( 'Email address input is empty.', 'wp_smaily' ) ), $redirect_url );
-		wp_safe_redirect( $redirect_url );
-		exit;
-	}
-
-
-
-	// Get data from database.
-	$table_name = esc_sql( $wpdb->prefix . 'smaily_config' );
-	$config     = $wpdb->get_row( "SELECT * FROM `$table_name`" );
-
-	// Make a opt-in request to server.
-	$server = 'https://' . $config->domain . '.sendsmaily.net/api/opt-in/';
-	$lang   = explode( '-', $params['lang'] );
-	$array  = array(
-		'remote'        => 1,
-		'success_url'   => $redirect_url,
-		'failure_url'   => $redirect_url,
-		'language'      => $lang[0],
-	);
-
-	// Add autoresponder if selected.
-	if ( (int) $config->autoresponder !== 0 ) {
-		$array['autoresponder'] = $config->autoresponder;
-	}
-
-	$form_values = [];
-	// Add custom form values to Api request if available.
-	foreach ( $params as $key => $value ) {
-		if ( array_key_exists( $key, $array ) ) {
-			continue;
-		} else {
-			$form_values[ $key ] = $value;
-		}
-	}
-
-	$array = array_merge( $array, $form_values );
-	require_once( SMLY4WP_PLUGIN_PATH . '/code/Request.php' );
-	$result  = (new Smaily_Plugin_Request())
-		->setUrl($server)
-		->setData($array)
-		->post();
-
-	if ( empty( $result ) ) {
-		$redirect_url = add_query_arg( 'smaily_form_error',
-			rawurlencode( __( 'Something went wrong', 'wp_smaily' ) ), $redirect_url );
-	} elseif ( (int) $result['code'] !== 101 ) {
-		switch ( (int) $result['code'] ) {
-			case 201:
-				$redirect_url = add_query_arg( 'smaily_form_error',
-					rawurlencode( __( 'Form was not submitted using POST method.', 'wp_smaily' ) ), $redirect_url );
-				break;
-			case 204:
-				$redirect_url = add_query_arg( 'smaily_form_error',
-					rawurlencode( __( 'Input does not contain a recognizable email address.', 'wp_smaily' ) ), $redirect_url );
-				break;
-			case 205:
-				$redirect_url = add_query_arg( 'smaily_form_error',
-					rawurlencode( __( 'Could not add to subscriber list for an unknown reason. Probably something in Smaily.', 'wp_smaily' ) ), $redirect_url );
-				break;
-			default:
-				$redirect_url = add_query_arg( 'smaily_form_error',
-					rawurlencode( __( 'Something went wrong', 'wp_smaily' ) ), $redirect_url );
-				break;
-		}
+/**
+ * Render notification bar.
+ *
+ * @return void
+ */
+function smaily_handle_response() {
```

Comment:
Good initiative, but the issue with this implementation is that the notification bar can break the layout of the site if it is injected at the wrong time (this conflict can happen with other plugins). Maybe it is better to display these errors next to the widget (you can expand the current `form_has_error` functionality).

Reply:
Then responses will be shown only when the user is redirected back or the form is on the page. The wp hook seems like a really high-level hook and might not be the right one for the job. What about a different hook?

Reply:
Perhaps there is a better hook for that. I hoped Wordpress had some sort of built-in notification/messaging system that we could have used (there is something for the admin, but not for public side). If you cannot find a better hook or mechanism for displaying the message, then I suggest you just show it inside the widget - it won't work as our other extensions/plugins/modulues do, but at least it is not breaking anything.

Reply:
Expanded the current form_has_error functionality but renamed it to form_has_response